infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java

```diff
@@ -230,9 +230,18 @@ private String getConflictHashCode(UUID streamId, byte[] conflictParam) {
     private TokenType txnCanCommit(TxResolutionInfo txInfo, /** Input. */
                                   AtomicReference<byte[]> conflictKey /** Output. */) {
         log.trace("Commit-req[{}]", txInfo);
-        final long txSnapshotTimestamp = txInfo.getSnapshotTimestamp();
+        final Token txSnapshotTimestamp = txInfo.getSnapshotTimestamp();
+
+        // A transaction can start with a timestamp issued from a previous
+        // epoch, so we need to reject transactions that don't have a snapshot
+        // timestamp with a different epoch than the this sequencer's epoch.
+        if (txSnapshotTimestamp.getEpoch() != sequencerEpoch) {
```

Comment:
Add a unit test.

Reply:
Is this true? Isn't this the txSnapshotTimestamp which could be on a different epoch than the sequencer.. we need another param that reflects the actual epoch by the start of the transaction...

Reply:
txSnapshotTimestamp is extracted from txInfo, which contains the timestamp the transaction acquired when it started.

Reply:
mmm maybe the fact that it is called in the code txSnapshotTimestamp is misleading? cause for a moment I thought this referred to the actual requested timestamp for the snapshot, but this actually does not go back to the sequencer..